cep-6.md

```diff
@@ -0,0 +1,215 @@
+
+The notification message will be in the JSON file format. This will allow us to not only store the message itself but 
+also metadata about the message, including information about how often the client should display the message (more on 
+this in the next section).
+
+Here's an example of the `notifications.json` file which will be stored in the root of the channel directory structure.
+
+```json
+{
+  "notifications": [
+    {
+      "id": "1cd1d8e5-d96c-42d1-9c29-e8120ad80823",
+      "message": "Here is an example message that will be displayed to users",
+      "level": "info",
+      "created_at": "2022-04-26T11:50:34+00:00",
+      "expiry": 604800
+    }
+  ]
+}
+```
+
+Detailed overview of the JSON fields:
+
+- **notifications** [Array] holds zero or more notifications that will be displayed to the client.
+  - **id** [String] unique ID for the message itself. UUIDs are preferred, but there is no required format.
+  - **message** [String] message that gets displayed to users.
+  - **level** [String] one of (info|warning|error). These will let our users know the category of the message
+    and will also allow the client to apply different formatting rules (e.g. text color).
+  - **created_at** [String] ISO 8601 formatted timestamp showing the creation time of the message.
+  - **expiry** [Number] a number specifying how long in seconds the message is valid for.
+
+### How often will these messages appear?
+
+How often the messages appear will be configurable by the channel owners and the client. This will be accomplished by 
+the expiry field in the `notifications.json` file itself, but the client will the have ultimate say over whether this
+message is displayed. We will provide clients with a setting to permanently disable these messages in their `.condarc`
+files:
+
+```yaml
+display_notifications: false
+```
+
+## Motivation
+
+The motivation for this CEP came about as channel owners (Anaconda specifically) needed a way to broadcast messages to
+users of their channels. These messages may contain specific notices for particular users (e.g. identified by
+IP address) or general messages for the wider audience of a particular channel.
+
+Additionally, this new notification space can also provide a place for us to relocate `conda update conda` reminders 
+to a more visible spot (at the end of command output versus in the middle of the output). On top of this, other channels
+can use these notifications as a way to share news with their users or "calls for help" in maintaining their channels.
+
+
+## Rationale
+
+TBD
+
+
+## Backwards Compatibility
+
+We do not expect any backwards compatibility issues for this new feature.
+
+
+## Alternatives
+
+- **Show notifications at the beginning of environment activation** This was deemed a little too intrusive/annoying.
+- **Show notifications at the beginning of command output** Users may miss this if placed here, especially for commands
+  with lots of output
```

Comment:
I'm not sure how many alternatives for message delivery we want to enumerate here, but for completeness, other options include:

* _Message in an HTTP header when retrieving any file from the repository_: This would be a better option for some kinds of messages, like download rate limiting or other blocks due to abuse, since it could be turned on by a rule in a CDN.  However, this use case is probably better addressed by having a standard way for conda to display errors on the console from HTTP status codes like 429 (Too Many Requests) and 403 (Forbidden).  Additionally, serving custom headers is challenging unless the repository owner has more control over the web server than is usually given by services like Github Pages.
* _Add notifications to the repodata.json file_: The repodata.json file is already being fetched, so adding some notifications would reduce the number of requests.  However, the repodata.json file is way too big already, so it would require clients to download a fairly large file before even being able to show the notification.
* _Create a generic metadata.json file, which contains a notifications key_: This is @wolfv 's suggestion, I think, and I do find it appealing for creating a space for other kinds of channel metadata.  The file should remain small, so it hopefully would not have the problems of putting notifications into repodata.json.  However, putting notifications into a generic metadata file would complicate dynamic generation of notifications (perhaps based on some client attribute, like IP address or something in the user agent) because the notifications would need to be combined with the other metadata.  For this reason, I think I still lean toward notifications being a separate file.

Reply:
Good thoughts. FYI there is a separate issue to make a better response to things like 403/429s. It was deemed that that (HTTP status codes) is more akin to better error reporting and something like this (channel notifications) would be better served under a different discussion.

Reply:
I added these to the CEP (I changed the wording a bit on the last one though).